cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go

```diff
@@ -458,3 +459,132 @@ func TestValidateFeatureGates(t *testing.T) {
 		}
 	}
 }
+
+func TestValidateArgSelection(t *testing.T) {
+	var tests = []struct {
+		name     string
+		c        *kubeadm.NodeConfiguration
+		expected bool
+	}{
+		{
+			"invalid: DiscoveryToken and DiscoveryFile cannot both be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "foo",
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryToken or DiscoveryFile must be set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:  "",
+				DiscoveryToken: "",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenAPIServers not set",
+			&kubeadm.NodeConfiguration{
+				DiscoveryToken: "abcdef.1234567890123456@foobar",
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
+		},
+		{
+			"invalid: DiscoveryTokenCACertHashes cannot be used with DiscoveryFile",
+			&kubeadm.NodeConfiguration{
+				DiscoveryFile:              "foo",
+				DiscoveryTokenCACertHashes: []string{"sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc"},
+			},
+			false,
```

Comment:
Do I read it correctly, those 3 test cases are exactly the same ?

Reply:
I forget why I put these three same test cases here, but I guess it was a stupid cope&paste mistake :sweat_smile:

Reply:
Please fix it :) We don't need 3 copies of the same test case.

Reply:
Yeah, will do :)